stream/stream_part.c

```diff
@@ -0,0 +1,180 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <libavutil/common.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "common/common.h"
+#include "options/m_option.h"
+#include "options/path.h"
+#include "stream.h"
+
+struct priv {
+    int64_t part_start;
+    int64_t part_max_end; // 0 for no limit
+    struct stream *inner;
+};
+
+static int fill_buffer(struct stream *s, void *buffer, int len)
+{
+    struct priv *p = s->priv;
+
+    if (p->part_max_end) {
+        // We don't simply use (s->pos >= size) to avoid early return
+        // if the file is still being appended to.
+        if (s->pos + p->part_start >= p->part_max_end)
+            return -1;
+        // Avoid rading beyond p->part_max_end
+        len = MPMIN(len, p->part_max_end - s->pos);
+    }
+
+    return stream_read_partial(p->inner, buffer, len);
+}
+
+static int seek(struct stream *s, int64_t newpos)
+{
+    struct priv *p = s->priv;
+    newpos += p->part_start;
+    return stream_seek(p->inner, newpos);
+}
+
+static int64_t get_size(struct stream *s)
+{
+    struct priv *p = s->priv;
+    int64_t size = stream_get_size(p->inner);
+    if (size <= 0)
+        return size;
+    if (size <= p->part_start)
+      return 0;
+    if (p->part_max_end)
+        size = MPMIN(size, p->part_max_end);
+    return size - p->part_start;
+}
+
+static void s_close(struct stream *s)
+{
+    struct priv *p = s->priv;
+    free_stream(p->inner);
+}
+
+static int parse_part_range(stream_t *stream)
+{
+    struct priv *p = stream->priv;
+
+    struct bstr b_url = bstr0(stream->url);
+    struct bstr proto_with_range, inner_url;
+
+    bool has_at = bstr_split_tok(b_url, "@", &proto_with_range, &inner_url);
+
+    if (!has_at) {
+        MP_ERR(stream, "Expected part://start[-end]@URL: '%s'\n", stream->url);
+        return STREAM_ERROR;
+    }
+
+    if (!inner_url.len) {
+        MP_ERR(stream, "URL expected to follow 'part://start[-end]@': '%s'.\n", stream->url);
+        return STREAM_ERROR;
+    }
+    stream->path = bstrto0(stream, inner_url);
+
+    mp_split_proto(proto_with_range, &proto_with_range);
+    struct bstr range = proto_with_range;
+
+    struct bstr start, end;
+
+    bool has_end = bstr_split_tok(range, "-", &start, &end);
+
+    if(!start.len) {
+        MP_ERR(stream, "The byte range must have a start, and it can't be negative: '%s'\n", stream->url);
+        return STREAM_ERROR;
+    }
+
+    if (has_end && !end.len) {
+        MP_ERR(stream, "The byte range end can be omitted, but it can't be empty: '%s'\n", stream->url);
+        return STREAM_ERROR;
+    }
+
+    const struct m_option opt = {
+        .type = &m_option_type_byte_size,
+        .min = 0,
+        .max = INT64_MAX,
```

Comment:
I think this can be removed, since the variable is int64_t in the first place, and this option parser does not allow negative values.